src/vs/workbench/services/configuration/browser/configurationService.ts

```diff
@@ -739,15 +743,35 @@ export class WorkspaceService extends Disposable implements IWorkbenchConfigurat
 		const changes = this.compareFolders(this.workspace.folders, workspaceFolders);
 		if (changes.added.length || changes.removed.length || changes.changed.length) {
 			this.workspace.folders = workspaceFolders;
-			const change = await this.onFoldersChanged();
-			this.triggerConfigurationChange(change, previous, ConfigurationTarget.WORKSPACE_FOLDER);
-			this._onDidChangeWorkspaceFolders.fire(changes);
+			this.checkFoldersTrustState(changes, async () => {
+				const change = await this.onFoldersChanged();
+				this.triggerConfigurationChange(change, previous, ConfigurationTarget.WORKSPACE_FOLDER);
+				this._onDidChangeWorkspaceFolders.fire(changes);
+			});
 		} else {
 			this.triggerConfigurationChange(change, previous, ConfigurationTarget.WORKSPACE);
 		}
 		this.updateUntrustedSettings();
 	}
 
+	private checkFoldersTrustState(changes: IWorkspaceFoldersChangeEvent, action: () => void): void {
+		// Workspace is untrusted
+		if (this.workspaceTrustState !== WorkspaceTrustState.Trusted) {
+			action();
+			return;
+		}
+
+		// Workspace is trusted
+		this._onWillChangeWorkspaceFolders.fire({
```

Comment:
Event name `_onWillChangeWorkspaceFolders` and when this event is triggered is not matching. I would recommend to remove the workspace trust semantics from the event triggering. Fire this event always and the callback checks for workspace trust.